### PR TITLE
Payments

### PR DESCRIPTION
**Pull request by KaviHarjani**
The payment in screener.in, is not working inside of Webview, as the redirection isn't taking place, any loophole around would result in screener getting banned from Playstore, so adding the razorpay official plugin here

### CODE REVIEW

File: lib/webview.dart [hunk 1]
```diff
@@ -16,7 +16,12 @@ class _ScreenerAppState extends State<ScreenerApp> {
   late WebViewController controller;
   final _razorpay = Razorpay();
   var options = {};
-  late bool isLoading = false;
+  late final String _screenerHomeUrl =
+      widget.debug ? "http://10.0.2.2:8000" : 'https://www.screener.in/';
+  late String paymentUrl = "'$_screenerHomeUrl/payment/capture/'";
```
Comment (manujagobind) [inline lib/webview.dart:21]: this will add / twice before payment. Remove / at the end of screener.in in the above line

**Comment by manujagobind** [comment]
Please add PR summary.

File: lib/webview.dart [hunk 2]
```diff
@@ -3,10 +3,10 @@ import 'package:webview_flutter/webview_flutter.dart';
 import 'package:url_launcher/url_launcher.dart';
 import 'package:razorpay_flutter/razorpay_flutter.dart';
 import 'dart:convert';
-import 'package:http/http.dart' as http;
 
 class ScreenerApp extends StatefulWidget {
-  const ScreenerApp({Key? key}) : super(key: key);
+  final debug;
```
Comment (manujagobind) [inline lib/webview.dart]: Add type.

File: lib/webview.dart [hunk 3]
```diff
@@ -31,31 +36,51 @@ class _ScreenerAppState extends State<ScreenerApp> {
...
+        "hiddenField.value = params[key];" +
+        "form.appendChild(hiddenField);}}document.body.appendChild(form);form.submit();}" +
+        "post($postUrl, $postParam, method=$requestMethod)");
   }
 
-  void _handlePaymentSuccess(PaymentSuccessResponse response) {
-    var url = Uri.parse("http://10.0.2.2:8000/payment/capture/");
-    Map body = {
-      'razorpay_payment_id': response.paymentId,
+  _paymentbody(paymentId) {
```
Comment (manujagobind) [inline lib/webview.dart]: Use camel case.

File: lib/webview.dart [hunk 4]
```diff
@@ -92,47 +117,43 @@ class _ScreenerAppState extends State<ScreenerApp> {
...
+            body: WebView(
+              initialUrl: _screenerHomeUrl,
+              javascriptMode: JavascriptMode.unrestricted,
+              userAgent: _proxyUserAgent,
+              onWebViewCreated: (controller) {
+                this.controller = controller;
+              },
+              zoomEnabled: false,
+              navigationDelegate: (NavigationRequest request) {
+                if (request.url.contains('premium')) {
```
Comment (manujagobind) [inline lib/webview.dart]: Prefer using double quotes to be inline with other strings in this file.

File: lib/webview.dart [hunk 5]
```diff
@@ -92,47 +117,43 @@ class _ScreenerAppState extends State<ScreenerApp> {
...
+                } else if (request.url.contains("google")) {
+                  return NavigationDecision.navigate;
+                } else {
+                  _launchURL(request.url);
+                  return NavigationDecision.prevent;
+                }
+              },
+              onPageFinished: (String url) async {
+                if (url.contains('premium')) {
+                  await controller.runJavascript("if (document.getElementById('razorpay-info'))" +
```
Comment (manujagobind) [inline lib/webview.dart]: JS can be formatted with more line breaks for better readability. Not sure if we should do this though as the file may get too large then.

File: lib/webview.dart [hunk 6]
```diff
@@ -3,10 +3,10 @@ import 'package:webview_flutter/webview_flutter.dart';
 import 'package:url_launcher/url_launcher.dart';
```
Comment (manujagobind) [inline lib/webview.dart:3]: Getting a warning that launch from this package is deprecated and asks to use launchURL instead.

File: lib/main.dart [hunk 7]
```diff
@@ -1,244 +1,19 @@
...
-    var cookies = await CookieManager().getCookies(url: url);
-    cookieString = "";
-    for (var cookie in cookies) {
-      cookieString += cookie.name + "=" + cookie.value;
-      if (cookies.indexOf(cookie) != cookies.length - 1) {
-        cookieString += "; ";
-      }
-    }
+      title: 'Screener',
+      home: Splash(debug: true),
```
Comment (manujagobind) [inline lib/main.dart:15]: Debug to be false in production.

File: lib/splash.dart [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  State<Splash> createState() => _SplashState();
+}
+
+int checkInt = 0;
+
+class _SplashState extends State<Splash> {
+  @override
+  void initState() {
+    super.initState();
+    Future<int> a = CheckInternet().checkInternetConnection();
```
Comment (manujagobind) [inline lib/splash.dart]: Let's use a meaningful variable name.

File: lib/splash.dart [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    a.then((value) {
+      if (value == 0) {
```
Comment (manujagobind) [inline lib/splash.dart]: Let's use a relevant variable name instead of just "value".

File: lib/splash.dart [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        setState(() {
+          checkInt = 0;
```
Comment (manujagobind) [inline lib/splash.dart:23]: Let's use a meaningful variable name.

File: lib/splash.dart [hunk 11]
```diff
@@ -0,0 +1,57 @@
+import 'package:flutter/material.dart';
+import './checkinternet.dart';
+import './webview.dart';
+
+class Splash extends StatefulWidget {
+  final debug;
+  Splash({Key? key, required this.debug}) : super(key: key);
```
Comment (manujagobind) [inline lib/splash.dart]: declare const.

File: lib/splash.dart [hunk 12]
```diff
@@ -0,0 +1,57 @@
+import 'package:flutter/material.dart';
+import './checkinternet.dart';
+import './webview.dart';
+
+class Splash extends StatefulWidget {
+  final debug;
```
Comment (manujagobind) [inline lib/splash.dart]: Add type.

**Review by manujagobind — COMMENTED** [review]
Love it!

File: lib/webview.dart (hunk 6 above)
Comment (KaviHarjani) [reply lib/webview.dart:3]: Was depreciated in the newer version, we were using the old one, upgraded it

File: lib/main.dart (hunk 7 above)
Comment (KaviHarjani) [reply lib/main.dart:15]: Debug is set to false to test payments in local env, while pushing will switch to false

File: lib/splash.dart (hunk 10 above)
Comment (KaviHarjani) [reply lib/splash.dart:23]: CheckInt, for check-internet-integer, not good?

File: lib/splash.dart [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+      if (internetStatus == 0) {
+        setState(() {
+          checkInt = 0;
+        });
+        ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
+          content: Text('No internet connection!'),
+        ));
+      } else {
+        setState(() {
+          checkInt = 1;
```
Comment (manujagobind) [inline lib/splash.dart:30]: How are we using checkInt?

File: lib/checkinternet.dart [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import 'package:connectivity_plus/connectivity_plus.dart';
+
```
Comment (manujagobind) [inline lib/checkinternet.dart:2]: Is the filename checkinternet.dart as per naming conventions?

File: lib/splash.dart (hunk 13 above)
Comment (KaviHarjani) [reply lib/splash.dart:30]: If there's no internet we show the users that there's no internet as a snack bar with screeners logo in the centre

Comment (manujagobind) [reply lib/splash.dart:30]: That we are anyway doing using snackbar. How are we using checkInt variable?

Comment (manujagobind) [reply lib/splash.dart:30]: Is checkInt variable required?

Comment (manujagobind) [reply lib/splash.dart:30]: I see we removed on the other PR.

Comment (manujagobind) [reply lib/splash.dart:30]: Let's remove here too and get it merged so that we resolve any conflicts on other PRs if any.

**Review by manujagobind — APPROVED** [review]

**Comment by manujagobind** [comment]
Yes, I see that's happening via the message we're setting. What's the
purpose of checkInt within setState? How's the value of checkInt used?

**Pull request merged**